src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -390,33 +418,62 @@ async def run_setup_py(
     exported_target = request.exported_target
 
-    owned_deps = await Get(OwnedDependencies, DependencyOwner(exported_target))
-    transitive_targets = await Get(TransitiveTargets, Addresses([exported_target.target.address]))
+    owned_deps, transitive_targets = await MultiGet(
+        Get(OwnedDependencies, DependencyOwner(exported_target)),
+        Get(TransitiveTargets, Addresses([exported_target.target.address])),
+    )
+
     # files() targets aren't owned by a single exported target - they aren't code, so
     # we allow them to be in multiple dists. This is helpful for, e.g., embedding
     # a standard license file in a dist.
     files_targets = (tgt for tgt in transitive_targets.closure if tgt.has_field(FilesSources))
     targets = Targets(itertools.chain((od.target for od in owned_deps), files_targets))
-    sources = await Get(SetupPySources, SetupPySourcesRequest(targets, py2=request.py2))
-    requirements = await Get(ExportedTargetRequirements, DependencyOwner(exported_target))
 
-    # Nest the sources under the src/ prefix.
-    src_digest = await Get(Digest, AddPrefix(sources.digest, CHROOT_SOURCE_ROOT))
+    sources, requirements = await MultiGet(
+        Get(SetupPySources, SetupPySourcesRequest(targets, py2=request.py2)),
+        Get(ExportedTargetRequirements, DependencyOwner(exported_target)),
+    )
+
+    # Generate the setup script.
+    setup_kwargs = await Get(
+        SetupPyKwargs, SetupPyKwargsRequest(exported_target, requirements, sources)
+    )
+    setup_py_content = SETUP_BOILERPLATE.format(
+        target_address_spec=exported_target.target.address.spec,
+        setup_kwargs_str=distutils_repr(setup_kwargs.kwargs),
+    ).encode()
+
+    files_to_create = [
+        FileContent("setup.py", setup_py_content),
+        FileContent("MANIFEST.in", b"include *.py"),
+    ]
+    extra_files_digest, src_digest = await MultiGet(
+        Get(Digest, CreateDigest(files_to_create)),
+        # Nest the sources under the src/ prefix.
+        Get(Digest, AddPrefix(sources.digest, CHROOT_SOURCE_ROOT)),
+    )
+    chroot_digest = await Get(Digest, MergeDigests((src_digest, extra_files_digest)))
+
+    return SetupPyChroot(chroot_digest, setup_kwargs)
 
-    target = exported_target.target
-    provides = exported_target.provides
 
-    # Generate the kwargs to the setup() call.
+@rule
+async def determine_setup_kwargs(request: SetupPyKwargsRequest) -> SetupPyKwargs:
+    provides = request.exported_target.provides
+    target = request.exported_target.target
+
     setup_kwargs = provides.setup_py_keywords.copy()
     setup_kwargs.update(
         {
             "package_dir": {"": CHROOT_SOURCE_ROOT},
-            "packages": sources.packages,
-            "namespace_packages": sources.namespace_packages,
-            "package_data": dict(sources.package_data),
-            "install_requires": tuple(requirements),
+            "packages": request.sources.packages,
```

Comment:
I think this logic is universal: we always want pants to set these values, regardless of any custom logic the user has to set other kwargs such as version.

So I don't think these belong in this rule, but in the outer rule that calls it. Although they probably should only set these values if the user hasn't set them in the inner rule.

This may also mean that the request object doesn't need the sources or requirements fields?

Reply:
This might make sense. Let's save it for the followup that actually adds the union, though. It wouldn't make sense to apply this change yet, as you still cannot give your own custom implementation.